src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,310 @@
+
+        return new ItemResource(stack);
+    }
+
+    /**
+     * <strong>Note:</strong> This cannot be called before your item is registered
+     *
+     * @throws IllegalStateException If the backing registry is unavailable or not yet ready.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    public static ItemResource of(Holder<Item> holder) {
+        return of(holder.value());
+    }
+
+    /**
+     * A wrapped {@link ItemStack} which must never be modified or exposed. This will be a size of 1 so that we can make use
+     * of the fact it is already an instance with a data component map.
+     */
+    private final ItemStack innerStack;
+
+    private ItemResource(ItemStack stack) {
+        this.innerStack = stack;
+    }
+
+    @Override
+    public Item value() {
+        return innerStack.getItem();
+    }
+
+    /**
+     * @return The {@link Item} of the resource from the inner {@link ItemStack}
+     */
+    public Item getItem() {
+        return value();
+    }
+
+    @Override
+    public Holder<Item> getHolder() {
+        return innerStack.getItemHolder();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return innerStack.isEmpty();
+    }
+
+    /**
+     * {@return true if the stack components and instance matches the inner stack's components and instance} Uses the {@link ItemStack#isSameItemSameComponents(ItemStack, ItemStack)} method for comparison.
+     *
+     * @param stack the item stack to check
+     */
+    public boolean is(ItemStack stack) {
+        return ItemStack.isSameItemSameComponents(stack, innerStack);
+    }
+
+    /**
+     * {@return true if the item instance matches the backing instance value}
+     *
+     * @param item the item to check
+     */
+    public boolean is(ItemLike item) {
+        return is(item.asItem());
+    }
+
+    /**
+     * Tests an {@link ItemStack} predicate with the inner stack.
+     *
+     * @param predicate Predicate to perform the test with
+     * @return {@code true} if the test passed
+     */
+    public boolean test(Predicate<ItemStack> predicate) {
+        return predicate.test(innerStack);
+    }
+
+    @Override
+    public boolean isComponentsPatchEmpty() {
+        return innerStack.isComponentsPatchEmpty();
+    }
+
+    @Override
+    public ItemResource withPatch(DataComponentPatch patch) {
+        if (isEmpty() || patch.isEmpty() || innerStack.getComponentsPatch().equals(patch))
+            return this;
+
+        ItemStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+
+        return ItemResource.of(stack);
+    }
+
+    @Override
+    public <D> ItemResource with(DataComponentType<D> type, D data) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.set(type, data);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
```

Comment:
Similar to `without`, I think we should just check if the data is the same as the value is currently, rather than comparing all the components...